src/cpu/tsc.cpp

```diff
@@ -0,0 +1,257 @@
+		return cached_result;
+	}
+
+	// Select CPU performance (cycles to MHz) map
+	auto getCyclesMhzMap = []() -> const CyclesMhzMap& {
+		if (CPU_ArchitectureType < ArchitectureType::Intel286) {
+			return Intel86_Map;
+		} else if (CPU_ArchitectureType < ArchitectureType::Intel386Slow) {
+			return Intel286_Map;
+		} else if (CPU_ArchitectureType < ArchitectureType::Intel486OldSlow) {
+			return Intel386_Map;
+		} else if (CPU_ArchitectureType < ArchitectureType::Pentium ||
+		           CPU_ArchitectureType == ArchitectureType::Mixed) {
+			return Intel486_Map;
+		} else if (CPU_ArchitectureType < ArchitectureType::PentiumMmx) {
+			return IntelPentium_Map;
+		} else {
+			return IntelPentiumMmx_Map;
+		}
+	};
+
+	const auto& cycles_mhz_map = getCyclesMhzMap();
+	assert(!cycles_mhz_map.empty());
+
+	// Helper lambda to get std::map iterator before the given one
+	auto getIterBefore =
+	        [&cycles_mhz_map](const CyclesMhzMap::const_iterator& iter_after) {
+		        auto iter1 = cycles_mhz_map.begin();
+		        auto iter2 = iter1;
+		        while (++iter1 != iter_after) {
+			        iter2 = iter1;
+		        }
+		        return iter2;
+	        };
+
+	// Calculate the coefficient to be used to convert cycles to MHz
+	const auto iter_upper = cycles_mhz_map.upper_bound(cycles);
+	if (iter_upper == cycles_mhz_map.end()) {
+		const auto iter_lower = getIterBefore(cycles_mhz_map.end());
+		coeff = iter_lower->second / iter_lower->first;
+	} else if (iter_upper == cycles_mhz_map.begin()) {
+		coeff = iter_upper->second / iter_upper->first;
+	} else {
+		const auto iter_lower = getIterBefore(iter_upper);
+
+		const auto distance_lower = cycles - iter_lower->first;
+		const auto distance_upper = iter_upper->first - cycles;
+
+		assert(distance_lower >= 0);
+		assert(distance_upper >= 0);
+
+		const auto range = distance_upper + distance_lower;
+
+		assert(range != 0);
+
+		const double coeff_lower = iter_lower->second / iter_lower->first;
+		const double coeff_upper = iter_lower->second / iter_upper->first;
+
+		coeff = (coeff_lower * distance_upper + coeff_upper * distance_lower) / range;
+	}
+
+	// Update the cache, return the result
+	cached_arch    = CPU_ArchitectureType;
+	cached_cycles  = cycles;
+	cached_result  = round_cpu_mhz(cycles * coeff);
+	is_cache_valid = true;
+
+	return cached_result;
+}
+
+static double get_estimated_cpu_mhz()
+{
+	if (CPU_CycleAutoAdjust) {
+		if (CPU_CycleLimit > 0) {
+			return get_estimated_cpu_mhz(CPU_CycleLimit);
+		} else {
+			// No desired number of cycles specified
+			return get_typical_cpu_mhz();
+		}
+	} else {
+		return get_estimated_cpu_mhz(CPU_CycleMax);
+	}
+}
+
+void CPU_ReadTSC()
+{
+	const auto cpu_mhz = get_estimated_cpu_mhz();
+
+	const auto tsc_precise = PIC_FullIndex() * cpu_mhz * 1000;
+	const auto tsc_rounded = static_cast<int64_t>(std::round(tsc_precise));
+
+	reg_edx = static_cast<uint32_t>(tsc_rounded >> 32);
+	reg_eax = static_cast<uint32_t>(tsc_rounded & 0xffffffff);
```

Comment:
Okay, I'll refrain from individual comments as my general thought is this should be vastly simplified.

1. TSC is Pentium only. Adding support for pre-Pentium CPUs muddies the waters and is not useful much, as I'll explain later. It just adds unnecessary complexity, so I suggest to simplify the whole thing.
2. Rounding to the nearest know CPU model is a questionable feature. I highly suggest to get rid of it. We can set any cycles number, so why round anything?  Also, you run Pentium CPUs on all sorts of speeds, depending on BIOS and motherboard support. For example, I'm underclocking my Pentium MMX 166 to 100 MHz because it runs cooler and slower is better for most DOS stuff anyway.
3. Ideally, you would get rid of the table approach and just have a simple formula that translates the cycles value to some TSC frequency, job done. Worst case, you could use a function with a. few linear segments to approximate the TSC for a given cycles value for `pentium` and `pentium_mmx` only. Then this whole file could go away and it will all become a one-liner or a small function somewhere else.

Keep it simple, don't add complexity, unless we really need it—that's my preference 😄

Reply:
@johnnovak I have simplified the code, there is no _round to the nearest real CPU clock_ code anymore, calculation segments were already working in a linear way.

@Torinde Pentium 200 MHz was really hard to get back then - believe me, I have tried to buy one in late 90s :) AFAIK the Pentium MMX over 233 MHz were mobile CPUs.

Reply:
Update: I have simplified it even further, there is a single coefficient to convert cycles to MHz for the Pentium, and another one for the Pentium MMX.

Reply:
Yeah that's what I hoped for when I made my above comment that something like this would be sufficient.

Reply:
Well, one could run few different benchmarks (Norton SysInfo, Quake, RAR, …) on a real CPU running at different frequencies, on a few different motherboards, compare this with DOSBox results, select some curve that looks similar, do the least square approximation to select the best curve parameters, outside of the normal frequency range use some linear extrapolation… doable (in fact, I did something similar at work once, to more-or-less predict how much time the calculations are going to take), but that would require a lot of work and a lot of vintage hardware. Definitely an overkill.